### DISCUSSION TITLE
Rnnning the Diag ROM on an M1 - issues

### DISCUSSION THREAD

**Discussion by neusse** [misterblack1/trs80-diagnosticrom#13] [Q&A]
I hope this is the correct place to discuss the use of the Diag ROM. 

Wanted to discuss some issues I am having and maybe some others are having as well.  

How I got here:
- I am the original owner of this Model 1.  Bought it new when I was 17 or 18.  
- I used Electric Pencil so I installed a lowercase mod.  My mod does not use a switch and the 8th bit seemed to be permeant. But it uses a spare gate and another signal to produce the select signals.  I can document it it needed. 
- I burned the 13 ROM on a 2716.  It verifies correctly after burning.
- The Model 1 has not been turned on in over 25 years.
- I had a bad Z80 and would not show signs of life.  It runs now after replacing the Z80.
- It is a 16K Model 1.
- It seems to have the 2 ROM level II Basic with the missing pin.  ver 1.2?
- I can run simple basic code no issue.

When trying to use the DIAG ROM 13:
- At first it would flash the screen testing the Video memory and then display a full screen of all the characters in several lines and hang.  Nothing out the cassette port.
- I have never been able to get any sound out of my system when running the rom.
- This made me wonder if the chip select going through a spare gate had some issue.  I removed the MOD and only wired the chip select lines to the piggy back coming from Z30.  No additional logic as I had before. This ran better in that it stopped displaying a screen of characters.
- Still no sound
- sometimes it will say it sees 8 bit video and hang.
- sometimes it will say testing video mem and hang
- sometimes it will say testing video me and testing 4k and hang.
- sometimes it only displays the banner and nothing else.
- It never starts testing the 16k of memory since I understand it will not since it thinks there is only 4 K.
- It is set correctly for 16K and Basic sees the correct amount when hitting enter on boot.

Just wondering what I can do to test the full amount of memory.  I also have an expansion interface I would like to test memory in.  But one thing at a time.

I do have pictures of the screen for everything I have described if need be.

Thanks for any help or any thoughts.

Regards,

George

**Comment by misterblack1** [comment]
Hi I'd suspect your EPROM is problematic due to the erratic behavior. 2716 chips require 25 volts to program properly and the TL866 is extremely bad at programming them. Try programming it 5 or even 10 times in a row and see if there is a difference... 

The ROM determines the RAM size by testing for RAM above 12k: "A machine with 4K bank size cannot have RAM at 12K ($7000). If the ROM tests that region and finds all bits bad, it assumes this is a 4K machine." (From the readme)

I would say it's next to impossible all of the RAM is bad at that address (all bits) so there are other issues with the machine. Since basic works, try poking and peeking values into location $7000 and see if it's indeed bad or working.

**Reply by neusse** [reply]
The 2716 I have required 21 volts.  And yes that is what was applied.  The ROM was also verified good on a compare read with the programmer. So I believe the ROM is good.  I will try the pokes at high memory.  

Also once I removed the lowercase MOD that used a spare gate so the DIAG ROM could recognize the Video RAM as 8bit correctly.  I have never seen the particular LC mod anywhere, But I do remember it had to do with Electric Pencil.  If anyone else has that LC mod the DIAG ROM will not work like mine.  The use of the gate on the control lines to the piggyback chip does not allow the correct timing for the test to be successful.  But I did remove it and tied it to the correct signals on z30.  

I will post some pics of what I was/am getting.

Thanks,

**Reply by ki3v** [reply]
If you're still seeing issues after removing the LC mod, that makes me wonder if there is a problem in some of the address decoding logic (chip selects), or maybe the bus transceivers?  Or the ribbon cable?

I've also heard that a weak or otherwise troublesome power supply can make TRS-80's very flaky.

I'm really a software guy, not really a hardware guy, so maybe I should shut up and leave it to the experts.

**Comment by neusse** [comment]
I have discovered some things that might be of interest.  At first, I thought my M1 had the Electric Pencil Lowercase Mod. It did not have that LC MOD.  My machine has the LC MOD from one of the Wizard books (The Custom TRS80 Complete).  It is a very nice MOD in that it does not require any switch.  I had removed the MOD and enabled the classic Electric Pencil Mod and things were not good in either switch setting. So I am reinstalling this Mod from the Wizard Book.

What I wanted to say here is that if the Wizard Mod is installed the Diagnostic ROM will fail. It will only display several lines of characters filling the screen and hang. I don't think it can pass the VRAM test ever with wizard LC MOD.

With all Mods removed I could get a 7bit detection from the ROM as well as an 8 Bit detection with the Electric Pencil Mod.  But in no case am I able to get any of the RAM tested.  hmmm.   Basic II does see a MEM of 15572 though and has no issues running or strange resets.  So time to move on.

But for anyone interested, there is a no-switch required LC Mod that works out in the world.




![Screenshot 2022-10-31 110325](https://user-images.githubusercontent.com/6461221/199078630-6f82dbc1-f511-4d64-b1f1-447fbd6b5865.png)
![Screenshot 2022-10-31 110221](https://user-images.githubusercontent.com/6461221/199078702-a25cd219-6ebd-4658-a981-3e324ef68e54.png)

**Reply by byteshift69** [reply]
I can also confirm that the machine will hang after the ASCII display, with the mod from this book installed.
